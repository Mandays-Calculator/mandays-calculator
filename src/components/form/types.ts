import type { ReactElement, ReactNode } from "react";

export type BaseInputProps<Type = object> = Type & {
  name?: string;
  label?: string | ReactNode;
  error?: boolean;
  helperText?: string | ReactNode;
  isFastField?: boolean;
  readOnly?: boolean;
};

export type WithInputControllerType =
  | "text"
  | "date-picker"
  | "dropzone"
  | "pin-code"
  | "autocomplete"
  | "checkbox-group"
  | "date-time-picker"
  | "number-input";

export type BaseInputPropsWithAdornment<Type = object> = Type & {
  name: string;
  label?: string;
  endAdornment?: ReactElement;
  readOnly?: boolean;
  placeholder?: string;
};

export interface WithInputControllerValueProp {
  value?: string;
  selected?: string;
  startDate?: string;
  endDate?: string;
}

type FormValues = {
  [key: string]: any;
};

export type FormErrors = {
  [key in keyof FormValues]?: string;
};
