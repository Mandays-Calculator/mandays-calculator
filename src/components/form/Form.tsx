import type { PropsWithChildren, ReactElement } from "react";
import type { FormikInstance } from "formik";
import { FormikProvider } from "formik";
interface FormProps<FormValues extends object> extends PropsWithChildren {
  instance: FormikInstance<FormValues>;
}

export const Form = <FormValues extends object>(props: FormProps<FormValues>): ReactElement => {
  const { instance, children, ...rest } = props;
  return (
    <FormikProvider value={instance}>
      <form
        {...rest}
        onSubmit={instance.handleSubmit}
      >
        {children}
      </form>
    </FormikProvider>
  );
};

export default Form;
