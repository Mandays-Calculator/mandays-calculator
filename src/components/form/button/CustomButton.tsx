import React from "react";
import Button, { ButtonProps } from "@mui/material/Button";
import { styled } from "@mui/material/styles";

type ColorVariants =
  | "primary"
  | "primaryDark"
  | "primaryLight"
  | "secondary"
  | "error"
  | "warning"
  | "success"
  | "neutral";

interface StyledButtonProps extends ButtonProps {
  colorVariant: ColorVariants;
}

const StyledButton = styled(Button)<StyledButtonProps>(
  ({ theme, colorVariant }) => {
    const colorsMap: Record<ColorVariants, string> = {
      primary: theme.palette.primary.main,
      primaryDark: theme.palette.primary.dark,
      primaryLight: theme.palette.primary.light,
      secondary: theme.palette.secondary.main,
      error: theme.palette.error.main,
      warning: theme.palette.warning.main,
      success: theme.palette.success.main,
      neutral: "#FFFFFF",
    };
    const contrastTextMap: Record<ColorVariants, string> = {
      primary: theme.palette.primary.contrastText,
      primaryDark: theme.palette.primary.contrastText,
      primaryLight: "#000000",
      secondary: theme.palette.primary.contrastText,
      error: theme.palette.primary.contrastText,
      warning: theme.palette.primary.contrastText,
      success: theme.palette.primary.contrastText,
      neutral: "#000000",
    };

    return {
      backgroundColor: colorsMap[colorVariant],
      color: contrastTextMap[colorVariant],
      textTransform: "none",
      "&:hover": {
        backgroundColor: colorsMap[colorVariant], // Keep the same background color on hover
        opacity: 0.9,
      },
    };
  }
);
interface CustomButtonProps extends ButtonProps {
  colorVariant?: ColorVariants;
}

export const CustomButton: React.FC<CustomButtonProps> = ({
  colorVariant = "primary",
  ...props
}) => {
  return <StyledButton colorVariant={colorVariant} {...props} />;
};

export default CustomButton;
