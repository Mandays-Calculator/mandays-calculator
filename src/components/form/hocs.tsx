import type { ComponentType, ReactElement } from "react";
import type { FormikValues } from "formik";
import type {
  BaseInputProps,
  WithInputControllerType,
  WithInputControllerValueProp,
} from "./types";
import type { ReactDatePickerFunctionParams } from "../global-types/app";

import { useMemo } from "react";
import { useFormikContext, getIn, Field, FastField } from "formik";
import { isArray } from "lodash";

type WrappedInputProps<Type> = Type &
  BaseInputProps & {
    name: string;
  };
type WithInputController<Type extends object> = (props: WrappedInputProps<Type>) => ReactElement;

export const withInputController = <Type extends FormikValues>(
  WrappedInput: ComponentType<Type>,
  inputType: WithInputControllerType = "text"
): WithInputController<Type> => {
  return (props: WrappedInputProps<Type>): ReactElement => {
    const { name, isFastField, ...rest } = props;
    const form = useFormikContext<Type>();

    const fieldValue = form.values[name] || getIn(form.values, name);
    const fieldError = form.errors[name] || getIn(form.errors, name);
    const fieldTouched = form.touched[name] || getIn(form.touched, name);
    const hasError = Boolean(fieldError) && Boolean(fieldTouched);

    const handleInputNumberChange = (_: any, value: number): void => {
      form.setFieldValue(name, value);
    };

    const handleDateChange = (date: ReactDatePickerFunctionParams): void => {
      form.setFieldValue(name, date);
    };

    const handleDropzoneChange = (files: File[]): void => {
      form.setFieldValue(name, files);
    };

    const handlePinCodeChange = (pinCode: string): void => {
      form.setFieldValue(name, pinCode);
    };

    const handleCheckBoxChanges = (checkBoxName: string, checked: boolean): void => {
      if (checked) {
        form.setFieldValue(name, [checkBoxName, ...(form.values[name] ?? [])]);
      } else {
        const checkBoxData: string[] = [...form.values[name]];
        const index = checkBoxData.indexOf(checkBoxName);

        if (index !== -1) {
          checkBoxData.splice(index, 1);
          form.setFieldValue(name, checkBoxData);
        }
      }
    };

    const handleAutoComplete = (data: Type): void => {
      form.setFieldValue(name, data);
    };

    const getInputChange = ():
      | typeof handleDateChange
      | typeof handleDropzoneChange
      | typeof handlePinCodeChange
      | typeof handleCheckBoxChanges
      | typeof handleAutoComplete
      | typeof handleInputNumberChange
      | typeof form.handleChange => {
      switch (inputType) {
        case "date-picker":
          return handleDateChange;
        case "dropzone":
          return handleDropzoneChange;
        case "pin-code":
          return handlePinCodeChange;
        case "checkbox-group":
          return handleCheckBoxChanges;
        case "autocomplete":
          return handleAutoComplete;
        case "number-input":
          return handleInputNumberChange;
        default:
          return form.handleChange(name);
      }
    };

    const getInputValueProp = (): WithInputControllerValueProp => {
      if (inputType === "date-picker") {
        if (isArray(fieldValue)) {
          return {
            startDate: fieldValue[0],
            endDate: fieldValue[1],
          };
        }
        return {
          selected: fieldValue,
        };
      }

      return {
        value: fieldValue,
      };
    };

    const MemoriedField = useMemo(() => (isFastField ? FastField : Field), [isFastField]);
    return (
      <MemoriedField
        component={WrappedInput}
        name={name}
        error={hasError}
        {...getInputValueProp()}
        onChange={getInputChange()}
        {...(rest as Type)}
      />
    );
  };
};
