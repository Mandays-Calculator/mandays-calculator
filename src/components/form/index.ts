export * from "./checkbox";
export * from "./checkbox-group";
export * from "./text-field";
export * from "./date-picker";
