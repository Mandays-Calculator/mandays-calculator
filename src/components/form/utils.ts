import type { FormErrors } from "./types";
import { defaultTo, get } from "lodash";

export const getOption = <Type extends object>(
  path: keyof Type | "value" | "label",
): ((option: Type) => string) => {
  return (option: Type) => {
    return defaultTo(get(option, path), "");
  };
};

export const getFieldError = (
  errors: FormErrors | undefined,
  name: string,
): string | undefined => {
  if (errors) {
    return typeof get(errors, name) === "string" ? get(errors, name) : "";
  }
};
