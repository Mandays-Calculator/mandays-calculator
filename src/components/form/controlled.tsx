import { CheckBox, CheckboxGroup, DatePicker, SearchSelect, TextField, NumberInput } from ".";
import DateTimePicker from "./date-time-picker/DateTimePicker";
import { withInputController } from "./hocs";
import Select from "./select/Select";
import TextArea from "./text-area/TextArea";

export const ControlledTextField = withInputController(TextField);
export const ControlledCheckBox = withInputController(CheckBox);
export const ControlledTextArea = withInputController(TextArea);
export const ControlledCheckBoxGroup = withInputController(
  CheckboxGroup,
  "checkbox-group"
) as typeof CheckboxGroup;
export const ControlledDatePicker = withInputController(
  DatePicker,
  "date-picker"
);
export const ControlledSelect = withInputController(Select);
export const ControlledSearchSelect = withInputController(
  SearchSelect,
  "autocomplete"
) as typeof SearchSelect;
export const ControlledDateTimePicker = withInputController(
  DateTimePicker,
  "date-time-picker"
);
export const ControlledNumberInput = withInputController(NumberInput, "number-input");
