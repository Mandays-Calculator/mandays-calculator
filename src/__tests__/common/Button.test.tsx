import { render } from "@testing-library/react";
import { CustomButton } from "~/components/form/button";

describe("CustomButton", () => {
  it("renders without crashing", () => {
    const { getByRole } = render(<CustomButton />);
    const button = getByRole("button");
    expect(button).toBeInTheDocument();
  });

  it("renders with different color variants", () => {
    const variants: any[] = [
      "primary",
      "primaryDark",
      "primaryLight",
      "secondary",
      "error",
      "warning",
      "success",
      "neutral",
    ];

    variants.forEach((variant) => {
      const { getByRole, unmount } = render(
        <CustomButton colorVariant={variant} />
      );
      const button = getByRole("button");
      // For this test, we're just checking if the button has some style when a variant is given
      // This doesn't validate the exact color, but ensures some styling has been applied
      expect(button).toHaveStyle("textTransform: none");
      unmount(); // unmount current variant before testing the next
    });
  });
});
