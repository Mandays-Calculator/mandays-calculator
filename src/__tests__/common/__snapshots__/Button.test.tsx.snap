// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomButton renders with different color variants 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u89a6x-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 2`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-t5cdwg-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 3`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-19575oc-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 4`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-bt8ywo-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 5`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-uq8efq-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 6`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-9vtvyq-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 7`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-h7qbqx-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders with different color variants 8`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1cv601t-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`CustomButton renders without crashing 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u89a6x-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
